mozfldp/server.py

```diff
@@ -24,74 +19,120 @@ class ServerFacade:
+        self._coef = coef
+        self._intercept = intercept
 
-        self._buffer = {"some_client_uuid_1": [], "some_client_uuid_2": []}
+        self._num_client = num_client
+        self._client_fraction = client_fraction
+
+        self._client_coefs = []
+        self._client_intercepts = []
+        self._num_samples = []
 
     def ingest_client_data(self, client_json):
-        # TODO: we need to read one sample of data from a client and
-        # add it to a sample buffer.  The intent here is that we want
-        # to collect enough data that we can safely sample from the
-        # buffer
-
-        # Assume that client_json is some JSON blob with everything we
-        # need to call client_update successfully.  We should have a
-        # `client_id` key that uniquely identifies the client so that
-        # we can pin data to a particular client.
-
-        # TODO: maybe something like this?
-        # self._buffer[client_json['client_id']].append(client_json)
-        pass
-
-    def update_classifier(self):
-        # TODO: this is where the magic happens
-        #
-        # I think roughly, this should be applying a sample of client
-        # data through a training loop.
-        #
-        # I believe you want to take the code from:
-        # https://github.com/mozilla/CANOSP-2019/blob/master/simulation_util.py#L98
-        # to
-        # https://github.com/mozilla/CANOSP-2019/blob/master/simulation_util.py#L149
-        # and update the parameters of the SGDClassifier.
-        #
-        # The samples `S` in simulation_util.py would be equivalent to self._buffer
-        # in this class
-        pass
-
-    def classify(self, some_input_json):
         """
-        Apply the classifer to some sample input
+        Accepts new weights from a client and stores them on the server side for averaging
+        
+        Args: 
+            client_json: a json object containing coefs, intercepts, and num_samples
+        """
+        client_json = json.loads(client_json)
+        self._client_coefs.append(client_json["coefs"])
+        self._client_intercepts.append(client_json["intercept"])
+        self._num_samples.append(client_json["num_samples"])
+
+    def compute_new_weights(self):
+        """ 
+        Applies the federated averaging on the stored client weights for this round
+        and return the new weights
         """
-        reshaped_X_test = None
-        reshaped_Y_test = None
-        score = self._classifier.score(reshaped_X_test, reshaped_Y_test)
-        return score
 
+        new_coefs = np.zeros(self._coef.shape, dtype=np.float64, order="C")
+        new_intercept = np.zeros(self._intercept.shape, dtype=np.float64, order="C")
 
-server = ServerFacade()
+        for index, (client_coef, client_intercept) in enumerate(
+            zip(self._client_coefs, self._client_intercepts)
+        ):
+            n_k = self._num_samples[index]
+            added_coef = [
```

Comment:
It might be simpler to `np.array` the `cleint_coef` and `client_intercept` beforehand, and then just use a single statement to multiply (using the vectorized operator).

Reply:
Forgot to push the commit before this PR got merged. Added this in #50